ext/php7/ip_extraction.c

```diff
@@ -0,0 +1,557 @@
+    }
+
+err:
+    efree(addr);
+    return res;
+}
+
+static bool _parse_ip_address_maybe_port_pair(const char *addr, size_t addr_len, ipaddr *out) {
+    if (addr_len == 0) {
+        return false;
+    }
+    if (addr[0] == '[') {  // ipv6
+        const char *pos_close = memchr(addr + 1, ']', addr_len - 1);
+        if (!pos_close) {
+            return false;
+        }
+        return _parse_ip_address(addr + 1, pos_close - (addr + 1), out);
+    }
+    const char *colon = memchr(addr, ':', addr_len);
+    if (colon) {
+        return _parse_ip_address(addr, colon - addr, out);
+    }
+
+    return _parse_ip_address(addr, addr_len, out);
+}
+
+#define CT_HTONL(x)                                                                          \
+    ((((x) >> 24) & 0x000000FFU) | (((x) >> 8) & 0x0000FF00U) | (((x) << 8) & 0x00FF0000U) | \
+     (((x) << 24) & 0xFF000000U))
+
+static bool _is_private_v4(const struct in_addr *addr) {
+    static const struct {
+        struct in_addr base;
+        struct in_addr mask;
+    } priv_ranges[] = {
+        {
+            .base.s_addr = CT_HTONL(0x0A000000U),  // 10.0.0.0
+            .mask.s_addr = CT_HTONL(0xFF000000U),  // 255.0.0.0
+        },
+        {
+            .base.s_addr = CT_HTONL(0xAC100000U),  // 172.16.0.0
+            .mask.s_addr = CT_HTONL(0xFFF00000U),  // 255.240.0.0
+        },
+        {
+            .base.s_addr = CT_HTONL(0xC0A80000U),  // 192.168.0.0
+            .mask.s_addr = CT_HTONL(0xFFFF0000U),  // 255.255.0.0
+        },
+        {
+            .base.s_addr = CT_HTONL(0x7F000000U),  // 127.0.0.0
+            .mask.s_addr = CT_HTONL(0xFF000000U),  // 255.0.0.0
+        },
+        {
+            .base.s_addr = CT_HTONL(0xA9FE0000U),  // 169.254.0.0
+            .mask.s_addr = CT_HTONL(0xFFFF0000U),  // 255.255.0.0
+        },
+    };
+
+    for (unsigned i = 0; i < ARRAY_SIZE(priv_ranges); i++) {
+        __auto_type range = priv_ranges[i];
+        if ((addr->s_addr & range.mask.s_addr) == range.base.s_addr) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool _is_private_v6(const struct in6_addr *addr) {
+    // clang-format off
+    static const struct {
+        union {
+            struct in6_addr base;
+            unsigned __int128 base_i;
+        };
+        union {
+            struct in6_addr mask;
+            unsigned __int128 mask_i;
+        };
+    } priv_ranges[] = {
+        {
+            .base.s6_addr = {[15] = 1},
+            .mask.s6_addr = {[0 ... 15] = 0xFF}, // /128
+        },
+        {
+            .base.s6_addr = {0xFE, 0x80},
+            .mask.s6_addr = {0xFF, 0xC0}, // /10
+        },
+        {
+            .base.s6_addr = {0xFC},
+            .mask.s6_addr = {0xFE}, // /7
+        },
+    };
+    // clang-format on
+
+    unsigned __int128 addr_i;
+    memcpy(&addr_i, addr->s6_addr, sizeof(addr_i));
+
+    for (unsigned i = 0; i < ARRAY_SIZE(priv_ranges); i++) {
+        __auto_type range = &priv_ranges[i];
```

Comment:
Can we avoid these GNU extensions here (__int128 and __auto_type)? If we ever want to port this to Windows: msvc does not support these.

Reply:
Done